### ISSUE TITLE
Warn on use of :or key from :as map in destructuring.

### ISSUE THREAD

**Issue by didibus** [clj-kondo/clj-kondo#1324]
**Is your feature request related to a problem? Please describe.**

```
(defn foo
  [{:keys [a] :as bar :or {a 10}}]
  [(:a bar) a])

(foo {})
;> [nil 10]

(foo {:a 1})
;> [1 1]
```
It can be surprising and lead to bugs that the `:as` bound map is the original input map, and it does not contain the `:or` key/value pairs in them.

**Describe the solution you'd like**

Which makes me think a linter that warned if someone tried to get one of the `:or` keys from the `:as` map would be a good reminder that they need to be sure of what they're doing, since it's not going to get the default value if missing.

**Describe alternatives you've considered**

None

**Additional context**

N/A

**Comment by borkdude** [comment]
I don't see how the linter could warn here without causing a false positive: the key a might as well be there in the `:as` map?

**Comment by jgomo3** [comment]
> I don't see how the linter could warn here without causing a false positive: the key a might as well be there in the `:as` map?

Yes. I think the work "might" is what the OP is concerned. That's why he talks about a warning as a "good reminder that they need to be sure of what they're doing".

But maybe that is not the job of a linter, but another similar tool.

**Comment by borkdude** [comment]
How is that different from `(defn foo [m] (:a m))` in general? clj-kondo doesn't warn about this like "hey, you're retrieving the value under key :a from map m, are you sure that's in there?".

**Comment by borkdude** [comment]
If there is no further argument to my last comment, I'll close this issue. Feel free to re-open if needed.

**Comment by didibus** [comment]
There is a difference, in your example the developers rightly expects it to return `nil` if it's not in there.

In my example, the developer wrongly expect it to return the default value from the `:or`.

The `:or` clause is most likely  a bug if the keys are retrieved from the `:as` map.

**Comment by borkdude** [comment]
I see the point now, thanks.

**Comment by borkdude** [comment]
So, to workaround the above problem I'm writing code like this:

```
(defn foo
  [{:keys [a] :as opts :or {a 10}}]
  ...
```
I would write in the body:

```
(let [opts (assoc opts :a a)] ...)
```

What would be a good heuristic that you're making a mistake?

**Comment by didibus** [comment]
Hum, at first I thought any use of the keyword instead of the symbol could be a warn, but I forgot about the assoc or similar case where you'd want to put the value back on something, you'd use the keyword then, and that would be a false positive.

Maybe just checking for common read patterns like:

`(:a opts)`
`(get opts :a)`
`(opts :a)`

would be good enough to catch most of this mistake, even though it wouldn't be exhaustive?
